File: python/problems/leetcode/0198_house_robber/solution.py
"""LeetCode 198: House Robber."""

from __future__ import annotations


class Solution:
    def rob(self, nums: list[int]) -> int:
        prev = curr = 0
        for x in nums:
            prev, curr = curr, max(curr, prev + x)
        return curr


if __name__ == "__main__":
    assert Solution().rob([1, 2, 3, 1]) == 4
    assert Solution().rob([2, 7, 9, 3, 1]) == 12
    assert Solution().rob([]) == 0
    print("0198 OK")

File: cpp/problems/leetcode/0198_house_robber/solution.cpp
// LeetCode 198
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int prev = 0, curr = 0;
        for (int x : nums) {
            int next = max(curr, prev + x);
            prev = curr;
            curr = next;
        }
        return curr;
    }
};

int main() {
    vector<int> a{1, 2, 3, 1};
    assert(Solution().rob(a) == 4);
    vector<int> b{2, 7, 9, 3, 1};
    assert(Solution().rob(b) == 12);
    vector<int> c{};
    assert(Solution().rob(c) == 0);
    cout << "0198 OK" << endl;
    return 0;
}
